### PR TITLE
Compiler support for respecting custom configured analyzer severity settings

### PR DESCRIPTION
**Pull request by mavasani**
Work towards #52991 and https://github.com/dotnet/roslyn/issues/27712#issuecomment-1794082012

See https://github.com/dotnet/roslyn/issues/52991#issuecomment-1709912198 for details. This is the first PR towards adding compiler layer functionality for this feature request.

This PR adds compiler layer support for a new `CustomSeverityConfigurable` [well-known diagnostic custom tag](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/Diagnostic/WellKnownDiagnosticTags.cs), such that:
   - An analyzer reporting at least one supported DiagnosticDescriptor with this custom tag is _always_ executed on all analyzer execution hosts.
   - An analyzer diagnostic reported with this custom tag is assumed to be configured to a custom severity with a custom syntax that is not known to the compiler. Compiler/AnalyzerDriver will not apply any additional severity configurations from editorconfig/globalconfig for such diagnostics.
 
There will be a few more follow-up PRs for completing this work. See https://github.com/dotnet/roslyn/issues/52991#issuecomment-1709912198 for details

### CODE REVIEW

File: src/Compilers/Core/Portable/Diagnostic/WellKnownDiagnosticTags.cs
Comment (mavasani) [inline src/Compilers/Core/Portable/Diagnostic/WellKnownDiagnosticTags.cs:51]: @sharwell This is the crux of the design that will ensure callbacks are always made into such analyzers, regardless of their default severity or any compiler-known severity configuration entries. We don't need the user project to have any dummy, blank source file with severities bumped to warning/error to ensure callbacks are always executed. The same tag is also used to ensure that `dotnet_diagnostic` and `dotnet_analyzer_diagnostic` entries are not applied to these diagnostics.

**Comment by mavasani** [comment]
Moving to draft until we have an API approval

**Comment by sharwell** [comment]
@mavasani I believe we can avoid most of this by just changing the default _declared_ severity of the IDE analyzers to warning, which removes the need to enable them manually.

**Comment by mavasani** [comment]
> @mavasani I believe we can avoid most of this by just changing the default _declared_ severity of the IDE analyzers to warning, which removes the need to enable them manually. The effective severity would not be altered by this change because it would continue to fall back to the customized severity for reporting. The only compile-side change we need is #70131.

@sharwell I don’t believe that will work. Compiler will override the severity from dotnet_diagnostic entries and bulk configuration entries. There is no way to prevent the compiler from doing so without communicating some information about the analyzer supporting custom configuration mechanisms.

**Comment by sharwell** [comment]
> Compiler will override the analyzer reported severity with the severity from dotnet_diagnostic entries and bulk configuration entries.

I'm not sure this is a problem though. We can mitigate this entirely from the .editorconfig editing experience without compiler changes, and no third-party analyzer will ever put together something this complicated because they were never forced into that box.

**Comment by mavasani** [comment]
> > Compiler will override the analyzer reported severity with the severity from dotnet_diagnostic entries and bulk configuration entries.
> 
> I'm not sure this is a problem though. We can mitigate this entirely from the .editorconfig editing experience without compiler changes, and no third-party analyzer will ever put together something this complicated because they were never forced into that box.

I am not sure I understand. Consider the following .editorconfig:

```editorconfig
csharp_style_var_for_built_in_types = true:warning
csharp_style_var_when_type_is_apparent = true:warning
csharp_style_var_elsewhere = true:warning

dotnet_diagnostic.IDE0007.severity = error
dotnet_analyzer_diagnostic.severity = error
```

An analyzer might report `IDE0007` diagnostics with severity `Warning` by decoding the first three entries, but if the editorconfig has one of the `dotnet_diagnostic` or `dotnet_analyzer_diagnostic` entries, they will override the analyzer reported severity with `Error` severity. How do you suggest preventing that without any compiler changes?

**Comment by mavasani** [comment]
> > How do you suggest preventing that without any compiler changes?
> 
> I don't suggest preventing that. I would instead suggest that we reveal the situation to the user:
> 
> 1. Fade out `:warning` as unnecessary code
> 2. Add a suggestion or warning for `:warning` saying the severity is being ignored because the severity `IDE0007` is being set separately
> 
> Users can then decide whether to remove the `:warning` or remove the explicit configuration of IDE0007. If the explicit severity of IDE0007 is removed, the warning for item (2) above could change to say the severity is ignored because the severity for an entire category is overriding it.

How would you handle that for bulk configuration entries? These are not conflicting ones, and still need to be skipped by the compiler

**Comment by mavasani** [comment]
> If the explicit severity of IDE0007 is removed, the warning for item (2) above could change to say the severity is ignored because the severity for an entire category is overriding it.

I don't believe this is actionable. User might want to bulk configure severity for an entire category or all analyzers + configure code style severities. The only option available with your proposal would be tell the user that they can use only one of these two features, which doesn't seem a nice experience, especially when both of them can be supported with the change in this PR.

**Comment by sharwell** [comment]
> The only option available with your proposal would be tell the user that they can use only one of these two features, which doesn't seem a nice experience, especially when both of them can be supported with the change in this PR.

This is fine, but I believe in the API review we should be clear that this feature is only needed for the specific case where the user wants to use bulk severity configuration of built-in code style analyzers _and_ further override that with line-item severity configurations.

A viable alternative to bulk configuration without a compiler change would be adding a new line item that built-in code style analyzers recognize that serves the same purpose as a bulk configuration but is self-applied by the analyzers instead of relying on the compiler for it. The editor experience could provide refactorings to automatically transition to this form if severity conflicts were detected.

Third-party analyzers do not need or benefit from these changes. If anything, it would make things more confusing if an analyzer started to declare support for custom severity only to leave users unable to figure out why the analyzer isn't respecting the built-in severity mechanisms.

**Comment by mavasani** [comment]
> > The only option available with your proposal would be tell the user that they can use only one of these two features, which doesn't seem a nice experience, especially when both of them can be supported with the change in this PR.
> 
> This is fine, but I believe in the API review we should be clear that this feature is only needed for the specific case where the user wants to use bulk severity configuration of built-in code style analyzers _and_ further override that with line-item severity configurations.
> 
> A viable alternative to bulk configuration without a compiler change would be adding a new line item that built-in code style analyzers recognize that serves the same purpose as a bulk configuration but is self-applied by the analyzers instead of relying on the compiler for it. The editor experience could provide refactorings to automatically transition to this form if severity conflicts were detected.
> 
> Third-party analyzers do not need or benefit from these changes. If anything, it would make things more confusing if an analyzer started to declare support for custom severity only to leave users unable to figure out why the analyzer isn't respecting the built-in severity mechanisms.

Personally, I feel there are lot of not-so-pleasant pieces with your proposal:
- All IDE code style analyzers will need to state they report warnings by default, when they actually do not. This means the rule nodes under `Dependencies -> Analyzers` will lie about the severity.
- We show conflicts warnings to the user between two supported severity configuration mechanisms `:severity` and `dotnet_diagnostic` ones, when we can easily detect and resolve these. Also note that `dotnet_diagnostic` entry might also be coming from a separate source, say a NuGet package that ships a globalconfig that has that entry, and the user is unable to modify that file.
- Bulk configuration cannot be supported

I believe all of these are tackled seamlessly with the change in this PR, which is introducing a very simple concept of `CustomConfigurable` analyzers and diagnostics, on the same lines as the existing `NotConfigurable` ones, and needs a very trivial public API addition.

**Comment by mavasani** [comment]
Public API was approved with https://github.com/dotnet/roslyn/issues/52991#issuecomment-1760359293. Marking ready for review.

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler for reviews

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs
Comment (cston) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs:453]: >foreach (var customTag in customTags) [](http://example.com/codeflow?start=12&length=37)

Not part of this change, but it looks like the loop can be replaced with `return customTags.Contains(tagToFind);`.

**Review by cston — APPROVED** [review]

**Comment by jaredpar** [comment]
@mavasani still waiting for feedback on this question 

https://github.com/dotnet/roslyn/issues/52991#issuecomment-1788123932

**Comment by mavasani** [comment]
> @mavasani still waiting for feedback on this question
> 
> [#52991 (comment)](https://github.com/dotnet/roslyn/issues/52991#issuecomment-1788123932)

Ah sorry, for some reason I missed that notification in my inbox. Your analysis is correct and I have added a comment to the issue with the possible approach to make this opt-in.

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler for second review.

**Comment by mavasani** [comment]
@333fred @jjonescz would one of you be able to review the PR?

Note that Jared's concern in the prior comment has been addressed with #70794 and https://github.com/dotnet/roslyn/pull/69844/commits/34bec2be13ecac199263777c968b31a92269ba38 pushed to #69844

**Review by jjonescz — APPROVED** [review]

**Comment by mavasani** [comment]
Thanks!

**Pull request merged**
